plugin/pkg/scheduler/algorithm/priorities/service_affinity.go

```diff
@@ -0,0 +1,148 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package priorities
+
+import (
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/plugin/pkg/scheduler/algorithm"
+)
+
+type ServiceAffinity struct {
+	serviceLister    algorithm.ServiceLister
+	controllerLister algorithm.ControllerLister
+}
+
+func NewServiceAffinityPriority(serviceLister algorithm.ServiceLister, controllerLister algorithm.ControllerLister) algorithm.PriorityFunction {
+	serviceAffinity := &ServiceAffinity{
+		serviceLister:    serviceLister,
+		controllerLister: controllerLister,
+	}
+	return serviceAffinity.CalculateAffinityPriority
+}
+
+// CalculateAffinityPriority affiliates a pod to the existing services' pods
+// that matches the labels indicated in affinitySelector.
+// The more existing pods to affiliate deployed on, the higher priority the node gets.
+func (s *ServiceAffinity) CalculateAffinityPriority(pod *api.Pod, podLister algorithm.PodLister, nodeLister algorithm.NodeLister) (algorithm.HostPriorityList, error) {
+	var maxCount int
+	counts := map[string]int{}
+	affinitySelector := labels.SelectorFromSet(labels.Set(pod.Spec.AffinitySelector))
+
+	// One direction: affiliate to pods of services
+	// 1) get services in the namespace same with pod
+	// 2) filter out services that matches affinitySelector
+	// 3) list services' pods to affiliate.
+	selectors := make([]labels.Selector, 0)
+	serviceList, err := s.serviceLister.List()
+	if err == nil {
+		for _, service := range serviceList.Items {
+			// consider only services that are in the same namespace as the pod
+			if service.Namespace != pod.Namespace {
+				continue
+			}
+			if affinitySelector.Matches(labels.Set(service.Labels)) {
```

Comment:
1. Using service.Labels means the pod logically affiliates to a given service (though it finally affiliates to pods of the service), no need to care about how the service find out its pods. (service.Spec.Selector is for service to find out pods belongs to itself. )

2. If use service.Spec.Selector, it means the user is clear how a given service find out its pods. In this case, matching service.Spec.Selector is duplicated here, since the final pods are directly shown by the user. We can match affinitySelector with existing pods' labels to filter out which to affiliate, as we did in [prior implementation](https://github.com/kubernetes/kubernetes/pull/14485#discussion-diff-40525708)


	
	```go
	serviceList, err := s.serviceLister.List()
	if err == nil {
		for _, service := range serviceList.Items {
			// consider only services that are in the same namespace as the pod
			if service.Namespace != pod.Namespace {
				continue
			}
			if affinitySelector.Matches(labels.Set(service.Spec.Selector)) {
				podSelector := labels.SelectorFromSet(labels.Set(service.Spec.Selector)))
				targetPods = append(targetPods, podLister.List(podSelector))
			}
		}
	}
	```
	can be optimized to
	```go
	targetPods := podLister.List(affinitySelector)

	```